specification/dartLangSpec.tex

```diff
@@ -16567,6 +16567,7 @@ \subsection{Identifier Reference}
 
 <typeIdentifier> ::= <IDENTIFIER>
   \alt <OTHER\_IDENTIFIER>
+  \alt \DYNAMIC{}
```

Comment:
Also needs `Function` (unless that's implicitly allowed as a special case of the function type syntax).

Reply:
As I mentioned, `type` [already derives](https://github.com/dart-lang/language/blob/ab893b49549bdd2bad750ca3a87fd0c309c21f85/specification/dartLangSpec.tex#L20047) `Function`. If `Function` is also added to `typeIdentifier` it is immediately an ambiguity.

@kaby76, why did you need to do that? Did you remove `Function` from the rule I linked above?

Reply:
Should `<typeNotVoidNotFunction>` really be named `<typeNotVoidNotFunctionNotDynamic>`?

Is there a good reason to treat `dynamic` and `Function` differently (say, because `?` is not allowed after `dynamic`?)

Reply:
`<typeNotVoidNotFunction>` is a type which is not the type `void` and not a function type (but it _does_ include `Function`). It does include `dynamic` as well: `<typeNotVoidNotFunction>` -> `<typeName>` -> `<typeIdentifier>` -> `dynamic`.

`Function` is tricky, because it's used in function types (`... Function(...)`), and also as a type in its own right.

Note that the spec grammar now says that `class dynamic {}` is OK, but the spec also says that [it is a compile-time error](https://github.com/dart-lang/language/blob/ab893b49549bdd2bad750ca3a87fd0c309c21f85/specification/dartLangSpec.tex#L16617) for a type declaration or a prefix to have a name which is a built-in identifier, so `class dynamic {}` is still an error after all.

But an extension could be named `dynamic` (e.g., `extension dynamic<X> on List<X> {...}`), which may or may not be a problem.